text/0730-semver-for-ts.md

```diff
@@ -0,0 +1,912 @@
+[emi-false]: https://www.typescriptlang.org/play?esModuleInterop=false&target=7#code/CYUwxgNghgTiAEBbA9sArhBByAzsxIAtGgA44AucUihEAlgEYywCeW8A3gLABQ8-8MMgB2FeADNkyAFzwAFAEp4AXgB88CjDrCA5gG5eA+CAAeJZDHIqJUgzwC+vXqEiwEKdJnhZELWo2YYNk5DATpEc0sbZAkYfG88AmIyShBqfyZWLDtQ-lNIqyFRK18AFQALbR1Zbj4jAUkZeHIWEhBkcWi7eoEGWFlFFXVhNEQGEBhugUcHIA
+
+Accordingly, library authors should set both `allowSyntheticDefaultImports` and `esModuleInterop` to `false`. This allows consumers to opt into these semantics, but does not *require* them to do so. Consumers can always safely use alternative import syntaxes (including falling back to `require()` and `import()`), or can enable these flags and opt into this behavior themselves.
+
+(If the Node ecosystem migrates full to ES modules over the next few years, this problem will be substantially mitigated.)
+
+
+### Design summary
+
+This section is a non-normative short summary for easy digestion. See the previous sections and [**Conformance**](#conformance) below for normative text.
+
+-   Public published types are part of the SemVer contract of a package, and must be versioned accordingly, per the specification above.
+
+-   Adding a new TypeScript version to the support matrix *may* cause breaking changes. When it does not, adding it is a normal minor release. When it *does*, cause a breaking change, the package must either mitigated that breakage (so consumers are not broken) *or* the package must release a major version.
+
+-   Removing a TypeScript version from the support matrix is a breaking change, except when it falls out of the supported version range under the “rolling support windows” policy.
+
+-   Packages using the “rolling window” policy should normally support all TypeScript versions released during the current ‘LTS’ of other core packages/runtimes/etc. they support, and drop support for them only when they drop support for that ‘LTS’, to minimize the number of major versions in the ecosystem.
+
+-   Both the currently-supported compiler versions and the compiler version support policy must be documented.
+
+-   Packages must be authored with the following compiler options:
+    -   `strict: true`
+    -   `noPropertyAccessFromIndexSignature: true`
+
+-   Libraries should generally be authored with the following compiler options:
+    -   `esModuleInterop: false`
+    -   `allowSyntheticDefaultImports: false`
+
+
+### Conformance
+
+To conform to this standard, a package must:
+
+- link to the final published version of this specification
+- specify the compiler support policy
+- specify the currently-supported versions of TypeScript
+- specify the definition of “public API” used by the library (e.g. “only documented types” vs. “all published types” etc.)
+- author and publish its types with `strict: true` and `noPropertyAccessFromIndexSignature: true` in its compiler configuration
+
+
+## How we teach this
```

Comment:
@emberjs/learning-team-managers definitely want to have the input of the Learning Core Team here. Not sure if this information should be a separate repository.

Reply:
Input very welcome!

The reason I suggested to put it in a separate repo is that the goal is (explicitly!) for it not to remain an Ember-specific policy and design, much as semver.org is used by many projects. We need this, for our adoption process; but many others would benefit from it as well.

Reply:
It seems to me that the intent here is to actually build a specification that will hopefully be adopted well beyond Ember. I would love to see this happen and it seems to me that it would be a huge blocker to this goal if the specification only lived within Ember’s documentation.

Reply:
It strikes me that another way this process could go is to develop the repo independent of Ember first then make an RFC for adopting this practices within Ember. However, it’s going to be important for the project to have at least one major adoptee and if it can’t even get Ember to adopt then it’s pretty much a non-starter. So to me, that makes a good case for doing it the way this RFC has proposed.

Reply:
I think it’s also worth noting that this spec is mainly for package developers not end users. All end users really want to know is “we won’t break your types willy-nilly.” Of course, having an actual spec with details that they can read is useful, but I strongly suspect that most users won’t really care about the details. If we do it right, things will just work. Where the spec matters a lot more is for those making a package who actually have to make sure they don’t break stuff! All that to say, I don’t think there really needs to be a detailed explanation on the main Ember docs beyond “we adhere to TS SemVer.”

Reply:
So, in terms of how we teach this, I like the idea of it being a separate repo when it's finished. Many successful specs are living documents that have changes over time, and I think it makes sense for those standards conversations to have a home that is not mixed in with conversation about other unrelated things (and vice versa). We could keep it as an RFC entry, but I think as a standalone it has a better chance of influencing the broader JS ecosystem to adopt the same things, versus creating their own RFC version of this with modifications.

An alternate approach is to make the separate repo now, and then link to it in this RFC as a proposal to adopt that document. That would make the Ember relationship a bit clearer, but I think the quality of feedback and iteration cycle really benefits from the full text being in the RFC. We just need to make sure there are clear, bold links to the other repo in this RFC later on.